simson/common/data_transformations.py

```diff
@@ -7,3 +9,60 @@ def broadcast_trailing_dimensions(array: np.ndarray, to_shape_of: np.ndarray) ->
     b_reshaped = np.reshape(array, new_shape)
     b_broadcast = np.broadcast_to(b_reshaped, to_shape_of.shape)
     return b_broadcast
+
+
+class Bound:
+    """Class representing bounds for a single parameter."""
+
+    def __init__(
+        self,
+        var_name: str,
+        lower_bound: Union[float, np.ndarray],
+        upper_bound: Union[float, np.ndarray],
+    ):
+        self.var_name = var_name
+        self.lower_bound = np.array(lower_bound)
+        self.upper_bound = np.array(upper_bound)
+
+    @staticmethod
+    def create_bounds_arr(
+        bounds_list: list["Bound"], all_prm_names: list[str], bound_shape: tuple
+    ) -> np.ndarray:
+        """Creates bounds array where each element is tuple of lower and upper bounds for each parameter."""
+
+        # Check bound shapes
+        if bounds_list:
+            if not bound_shape == bounds_list[0].lower_bound.shape:
+                raise ValueError("Bounds shape must match target shape")
+
+        if any(
+            b.lower_bound.shape != bound_shape or b.upper_bound.shape != bound_shape
+            for b in bounds_list
+        ):
+            raise ValueError("All bounds must have the same shape")
+
+        # Check for invalid parameter names
+        invalid_params = set(b.var_name for b in bounds_list) - set(all_prm_names)
+        if invalid_params:
+            raise ValueError(f"Unknown parameters in bounds: {invalid_params}")
+
+        bounds = np.empty(bound_shape, dtype=object)
+        param_positions = {name: i for i, name in enumerate(all_prm_names)}
+        for index in np.ndindex(bound_shape):
```

Comment:
That loops seems quite inefficient. 
I think some array-based code would be better than nested loops and conditions.